zplugin-ros1/src/ros_to_zenoh_bridge/ros1_to_zenoh_bridge_impl.rs

```diff
@@ -0,0 +1,233 @@
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+use async_std::sync::{Mutex, MutexGuard};
+use futures::{pin_mut, select, FutureExt};
+
+use zenoh;
+
+use std::sync::{
+    atomic::{AtomicBool, Ordering::Relaxed},
+    Arc,
+};
+
+use log::{debug, error};
+
+use crate::ros_to_zenoh_bridge::{
+    bridges_storage::BridgesStorage, discovery::LocalResources, environment::Environment,
+    ros1_client, topic_mapping, zenoh_client,
+};
+
+use super::{discovery::DiscoveryBuilder, ros1_client::Ros1Client};
+
+#[derive(PartialEq, Clone, Copy)]
+pub enum RosStatus {
+    Unknown,
+    Synchronizing,
+    Ok,
+    Error,
+}
+
+#[derive(Default, PartialEq, Eq, Clone, Copy)]
+pub struct BridgeStatus {
+    pub ros_publishers: (usize, usize),
+    pub ros_subscribers: (usize, usize),
+    pub ros_services: (usize, usize),
+    pub ros_clients: (usize, usize),
+}
+
+pub async fn work_cycle<RosStatusCallback, BridgeStatisticsCallback>(
+    session: Arc<zenoh::Session>,
+    flag: Arc<AtomicBool>,
+    ros_status_callback: RosStatusCallback,
+    statistics_callback: BridgeStatisticsCallback,
+) where
+    RosStatusCallback: Fn(RosStatus),
+    BridgeStatisticsCallback: Fn(BridgeStatus),
+{
+    let ros1_client = Arc::new(ros1_client::Ros1Client::new(
+        Environment::ros_name().get::<String>().as_str(),
+    ));
+    let zenoh_client = Arc::new(zenoh_client::ZenohClient::new(session.clone()));
+
+    let local_resources = Arc::new(LocalResources::new(
+        "*".to_string(),
+        "*".to_string(),
+        session.clone(),
+    ));
+
+    let bridges = Arc::new(Mutex::new(BridgesStorage::new(
+        ros1_client.clone(),
+        zenoh_client,
+        local_resources,
+    )));
+
+    let mut bridge = RosToZenohBridge::new(ros_status_callback, statistics_callback);
+
+    let discovery = make_discovery(session.clone(), bridges.clone()).fuse();
+
+    let run = bridge.run(ros1_client, bridges, flag).fuse();
+
+    pin_mut!(discovery, run);
+
+    select! {
+        _ = discovery => {},
+        _ = run => {}
+    };
+}
+
+async fn make_discovery<'a>(session: Arc<zenoh::Session>, bridges: Arc<Mutex<BridgesStorage>>) {
+    let mut builder = DiscoveryBuilder::new("*".to_string(), "*".to_string(), session);
+
+    let (add_tx, add_rx) = flume::unbounded();
+    let (rem_tx, rem_rx) = flume::unbounded();
+
+    let add_tx = Arc::new(add_tx);
+    let rem_tx = Arc::new(rem_tx);
+
+    builder
+        .on_discovered(move |b_type, topic| {
+            let add_tx = add_tx.clone();
+            Box::new(Box::pin(async move {
+                match add_tx.send_async((b_type, topic)).await {
```

Comment:
The `match` could be replaced with `if let Err(e) { .. }`. By doing this, there is no need to write `Ok(_) => {}`.